pkg/services/publicdashboards/commands/generate_datasources/generate.go

```diff
@@ -0,0 +1,103 @@
+package generate_datasources
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+	"sync"
+)
+
+type listPluginResponse struct {
+	Items []struct {
+		Slug string `json:"slug"`
+	} `json:"items"`
+}
+
+type datasourceDetails struct {
+	Slug     string `json:"slug"`
+	Backend  bool   `json:"backend"`
+	Alerting bool   `json:"alerting"`
+}
+
+type datasource struct {
+	Json datasourceDetails `json:"json"`
+}
+
+// Get the slugs of all datasource plugins that are compatible with public dashboards
+func GetCompatibleDatasources(baseUrl string) ([]string, error) {
+	slugs, err := getDatasourcePluginSlugs(baseUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	datasources := getDatasourceDetails(slugs, baseUrl)
+
+	// we only consider a datasource to be supported when alerting and backend are both true
+	var supported []string
+	for _, datasource := range datasources {
+		if datasource.Alerting && datasource.Backend {
+			supported = append(supported, datasource.Slug)
+		}
+	}
+
+	sort.Strings(supported)
+	return supported, nil
+}
+
+// Get a list of all the datasource plugin slugs
+func getDatasourcePluginSlugs(baseUrl string) ([]string, error) {
+	resp, err := http.Get(baseUrl + allPluginsEndpoint())
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	res := &listPluginResponse{}
+	err = json.NewDecoder(resp.Body).Decode(res)
+	if err != nil {
+		return nil, err
+	}
+	var slugs []string
+	for _, meta := range res.Items {
+		slugs = append(slugs, meta.Slug)
+	}
+	return slugs, nil
+}
+
+// Get the details for each datasource plugin by its slug
+func getDatasourceDetails(slugs []string, baseUrl string) []datasourceDetails {
+	datasources := make([]datasourceDetails, len(slugs))
+	var wg sync.WaitGroup
+	for i, slug := range slugs {
+		wg.Add(1)
+		// create new goroutine for each outgoing request
+		go func(i int, slug string) {
+			defer wg.Done()
+			url := baseUrl + pluginEndpoint(slug)
+			r, err := http.Get(url)
+			if err != nil {
+				panic(err)
+			}
+
+			datasource := &datasource{}
+			err = json.NewDecoder(r.Body).Decode(datasource)
+			if err != nil {
+				panic(err)
+			}
+			datasource.Json.Slug = slug
+			datasources[i] = datasource.Json
+		}(i, slug)
+	}
+
+	wg.Wait()
+
+	return datasources
+}
+
+func pluginEndpoint(slug string) string {
+	return fmt.Sprintf("/api/plugins/%s?version=latest", slug)
+}
+
+func allPluginsEndpoint() string {
+	return "/api/plugins?orderBy=weight&direction=asc&typeCodeIn[]=datasource"
+}
```

Comment:
do we need it as a function? why not to be just a const value

Reply:
It could be a constant. But since `pluginEndpoint()` needs to be a function I opted to keep them together for readability/consistency. I don't think it matters either way.

Reply:
greate, I agree